### ISSUE TITLE
Kendall's W is sometimes negative

### ISSUE THREAD

**Issue by IndrajeetPatil** [easystats/effectsize#448]
When it should always be in [0, 1].

``` r
library(effectsize)
library(statsExpressions) # for bugs dataset in tidy format

kendalls_w(desire ~ condition | subject, data = bugs_long, verbose = FALSE)
#> Kendall's W |        95% CI
#> ---------------------------
#> -0.35       | [-0.75, 1.00]
#> 
#> - One-sided CIs: upper bound fixed at [1.00].
```

<sup>Created on 2022-05-26 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by mattansb** [comment]
This is not related to missing values:

``` r
library(effectsize)
library(statsExpressions) # for bugs dataset in tidy format
#> Warning: package 'statsExpressions' was built under R version 4.1.3

kendalls_w(desire ~ condition | subject, data = na.omit(bugs_long), verbose = FALSE)
#> Kendall's W |        95% CI
#> ---------------------------
#> -0.37       | [-0.72, 1.00]
#> 
#> - One-sided CIs: upper bound fixed at [1.00].
```

<sup>Created on 2022-05-26 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

It's related to the way W is computed when there are ties, that seems to be able to be negative???

I'll have to take a closer look...

**Comment by mattansb** [comment]
Okay, I had a mistake in the correction for ties, but.... It's still not working :(

``` r
library(effectsize)
library(statsExpressions) # for bugs dataset in tidy format
#> Warning: package 'statsExpressions' was built under R version 4.1.3

kendalls_w(desire ~ condition | subject, data = bugs_long, verbose = FALSE)
#> Kendall's W |        95% CI
#> ---------------------------
#> -0.47       | [-1.29, 1.00]
#> 
#> - One-sided CIs: upper bound fixed at [1.00].
```

<sup>Created on 2022-05-27 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>


Can someone take a look here
https://github.com/easystats/effectsize/blob/fix-Ken-W/R/rank_effectsizes.R#L479-L508

and see what I'm doing wrong with regards to https://www.wikiwand.com/en/Kendall%27s_W#/Correction_for_Ties

**Comment by mattansb** [comment]
Looking at the implementation from `DescToosl`, it seems they are using a different formula than that one on Wiki (instead of `(12 * sum(R^2) - 3 * (m^2) * n * ((n + 1)^2))` in denominator, they use the standard `12 * S`):

https://github.com/AndriSignorell/DescTools/blob/5e676954da609a3531cf585c88e2ba5281c658a0/R/StatsAndCIs.r#L5681

It gives more sensible results, but are they the correct ones?

**Comment by etiennebacher** [comment]
But I think there's a mistake in the code in "fix-Ken-W" branch:
```r
Ti <- numeric(nrow(have_ties))
for (j in seq_len(nrow(have_ties))) {
  gs <- have_ties[j,][duplicated(have_ties[j,])]
  for (g in seq_along(gs)) {
    .t <- sum(have_ties[j,] == gs[g])
    Ti[j] <- (.t^3 - .t) + Ti[j]
  }
}
```
If there's a row with only 1s in `rankings`, this code outputs the wrong result:
``` r
have_ties <- matrix(rep(1, 4), nrow = 1)
have_ties
#>      [,1] [,2] [,3] [,4]
#> [1,]    1    1    1    1

Ti <- numeric(nrow(have_ties))
for (j in seq_len(nrow(have_ties))) {
  gs <- have_ties[j,][duplicated(have_ties[j,])]
  for (g in seq_along(gs)) {
    .t <- sum(have_ties[j,] == gs[g])
    Ti[j] <- (.t^3 - .t) + Ti[j]
  }
}
Ti
#> [1] 180
```

<sup>Created on 2022-05-31 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>


It should return 4^3 - 4 = 64 instead of 180.

**Comment by mattansb** [comment]
This last commit should fix that. But still the values are off due to that additional term in the denominator...

``` r
effectsize::kendalls_w(
  desire ~ condition | subject, 
  data = statsExpressions::bugs_long, 
  verbose = FALSE
)
#> Kendall's W |        95% CI
#> ---------------------------
#> -0.35       | [-0.81, 1.00]
#> 
#> - One-sided CIs: upper bound fixed at [1.00].
```

<sup>Created on 2022-06-01 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by mattansb** [comment]
![7f1d28957fb89a290a4b5a696114038f](https://user-images.githubusercontent.com/35330040/172544896-e16d1fbd-33fe-4094-9bd5-bc0785880e64.gif)
